Implement a RDMA software protocol stack

ruiqurm opened the pull request.

EAimTY committed:
Add control ringbuf descriptor types

EAimTY committed:
Add work ringbuf descriptor types

EAimTY committed:
Update desc fields

EAimTY committed:
Expose descriptor fields

EAimTY committed:
Make device initialization fallible

EAimTY committed:
derive `Debug`, `Clone` for several enums

EAimTY committed:
implement control path verbs

EAimTY committed:
expose public structs under `qp` module

EAimTY committed:
check associated QP before destroying PD

EAimTY committed:
implement send and recv interface

EAimTY committed:
implement polling logic

EAimTY committed:
fix send report handling

EAimTY committed:
add packet receiving completion check

EAimTY committed:
use global incremental id as ctrl op id

EAimTY committed:
hide magic numbers behind constants

EAimTY committed:
correctly calculate MR key

EAimTY committed:
fix `QP_AVAILABLITY` initialization

EAimTY committed:
fix incorrect ctrl descriptor content

EAimTY committed:
implement page table

EAimTY committed:
rename send and recv function

EAimTY committed:
reply ACK message

EAimTY committed:
add device type `software`

EAimTY committed:
open shared memory in emulated device

myrfy001 committed:
add rpc client and shared memory allocator

EAimTY committed:
simplify emulator rpc client

EAimTY committed:
move device constants to parent module

EAimTY committed:
add new descriptor abstraction

EAimTY committed:
refactor ringbuf and move it to the parent module

EAimTY committed:
adapt new descriptor types

EAimTY committed:
control path descriptor serialization / deserialization

EAimTY committed:
read / write control ringbuf on device

EAimTY committed:
add example `emulated`

EAimTY committed:
implement write for `ToCardWorkRbDesc`

EAimTY committed:
implement read for `ToHostWorkRbDesc`

EAimTY committed:
implement push and pop for work ringbuf in emulated device

EAimTY committed:
aadd `get_phys_addr` in trait `DeviceAdaptor`

EAimTY committed:
fix address calc for page table descriptor

EAimTY committed:
implement ringbuf interface directly on `EmulatedDevice`

EAimTY committed:
update rpc server addr in emulated device example

EAimTY committed:
remove dead code

myrfy001 committed:
fix bug and MR/QP create call works with emulator

EAimTY committed:
implement verb-level write API

rogercloud commented on src/device/constants.rs:
```diff
@@ -0,0 +1,54 @@
+#![allow(unused)]
```
Generally we don't use module level `allow`.

rogercloud commented on src/device/emulated/mod.rs:
```diff
@@ -1,32 +1,63 @@
+#![allow(unused)]
```
the same as above

rogercloud commented on src/device/emulated/mod.rs:
```diff
@@ -1,32 +1,63 @@
+    ) -> Result<Self, Box<dyn Error>> {
+        let rpc_client = Arc::new(RpcClient::new(rpc_server_addr, heap_mem_start_addr)?);
+
+        #[allow(clippy::arc_with_non_send_sync)]
```
Please leave a comment to explain the reason of allowing the lint.

rogercloud commented on src/device/emulated/mod.rs:
```diff
@@ -47,35 +78,55 @@ impl DeviceAdaptor for EmulatedDevice {
+            let p: *const ToHostCtrlRbDesc = a.as_ptr();
+            std::mem::transmute_copy(&*p)
+        }
+        // todo!()
```
Just remove the useless comment.

rogercloud commented on src/device/emulated/mod.rs:
```diff
@@ -47,35 +78,55 @@ impl DeviceAdaptor for EmulatedDevice {
+            let p = a.as_ptr();
+            std::mem::transmute_copy(&*p)
+        }
+        // todo!()
```
The same as above

rogercloud commented on src/device/emulated/ringbuf.rs:
```diff
@@ -0,0 +1,122 @@
+    pub(crate) fn new() -> Self {
+        let ringbuf_size_in_byte = D * E;
+        if ringbuf_size_in_byte < RINGBUF_PAGE_SIZE {
+            panic!("invalid ringbuf size.");
```
Please be careful with `panic`, if it's a recoverable error, use `Error` instead.

rogercloud reviewed (commented):
Please add clippy check and format ci in the github workflow. Datenlord project is a good example. We usually add more strict lint checking than the default setting, please check Datenlord's `main.rs` for reference.

myrfy001 committed:
fix more bug and can send work queue entry now

myrfy001 committed:
add some comment

myrfy001 committed:
fix bug and simple Write request can work now

EAimTY committed:
optimize ringbuf structure

EAimTY committed:
support RDMA read operation

myrfy001 committed:
read+write now works

myrfy001 committed:
refacetor example to support two nic for debug

ruiqurm committed:
squash into one commit

ruiqurm committed:
update descriptor

ruiqurm committed:
remove .vscode

ruiqurm committed:
use `bitfield` to refactor descriptor writing and reading

ruiqurm committed:
remove some `Copy` Trait

ruiqurm committed:
refactor: add the `responser` thread

the `responser` thread is used to sent the `ack`, `nack` and `read response` automaticlly
